### DISCUSSION TITLE
VirtIO error and then vm is stuck

### DISCUSSION THREAD

**Discussion by muhamadazmy** [cloud-hypervisor/cloud-hypervisor#2434] [Q&A] [answered]
Version: 0.13.0 (static build)
Kernel: 5.4.10

I use a RAW disk with my machine, it's an Fallocated disk (50G) over a physicall SSD disk (btrfs).  COW is disabled for the vdisk attached to the VM. 

```
cloud-hypervisor: 35.185492041s: <_disk0_q0> ERROR:virtio-devices/src/block.rs:271 -- Failed to process queue (submit): RequestParsing(DescriptorChainTooShort)
```

We **occasionally** get this error in CH logs, once this happens machine is not responsive anymore (still answer pings, ssh hangs as per the screen shot

![photo_2021-03-25_14-33-59](https://user-images.githubusercontent.com/10920323/112481400-439e3b00-8d77-11eb-8a27-5497b23100e6.jpg)

force restarting the machine (killing the CH process and start it over) fixes the problem. So we think it's not the issue with the vdisk or the physical disk themselves.

**Comment by sboeuf** [comment, marked as answer]
@muhamadazmy could you try to reproduce the error with latest version from master branch. It includes some additional error logs which might help us understand what is going on.

**Reply by muhamadazmy** [reply]
Sure, will do

**Reply by muhamadazmy** [reply]
Sorry it took a while for the issue to occur:
Please find below the only 2 log lines in the vm logs

```
cloud-hypervisor: 152.997984465s: <_disk0_q0> ERROR:block_util/src/lib.rs:198 -- Only head descriptor present: request = Request { request_type: Out, sector: 2276800, data_descriptors: [], status_addr: GuestAddress(0), writeback: true }
cloud-hypervisor: 152.998235295s: <_disk0_q0> ERROR:virtio-devices/src/block.rs:295 -- Failed to process queue (submit): RequestParsing(DescriptorChainTooShort)
```

**Reply by delandtj** [reply]
@sboeuf just a side-note: I wish I was living where you live. ;-)

**Reply by muhamadazmy** [reply]
So @sboeuf are the logs helpful ? do you need more debugging information ?

**Reply by sboeuf** [reply]
They are, but we're wondering why the descriptors are empty. I'm trying to find a way to reproduce on my side, as this will be easier to debug/fix.

**Reply by sboeuf** [reply]
In the meantime, if you get some spare time to help debug this issue, you could try git bisect to identify which commit introduced this problem.
Actually the first thing before doing git bisect is to find out if there's a version where this bug wasn't happening.

**Reply by sboeuf** [reply]
@muhamadazmy could you also provide the CH command line that you used?

**Reply by muhamadazmy** [reply]
> Actually the first thing before doing git bisect is to find out if there's a version where this bug wasn't happening.

Well, we started already with version 0.13.0 so we don't know if it worked on older version. Switching to use CH is recent, and we didn't face an issue during the initial testing.

The command line
```
cloud-hypervisor \
   --kernel /var/cache/modules/flistd/mountpoint/k8s:3f8760bf9e5c5e1caa907ebe6807bcf6/k3os-vmlinux \
   --initramfs /var/cache/modules/flistd/mountpoint/k8s:3f8760bf9e5c5e1caa907ebe6807bcf6/k3os-initrd-amd64 \
   --cmdline 'console=ttyS0 reboot=k panic=1 net_eth0=10.200.4.212/24,10.200.4.1,10.200.0.0/16,fd52:4d64:6d34:4::d4/64,fd52:4d64:6d34:4::1,priv net_dns=8.8.8.8,1.1.1.1,2001:4860:4860::8888' \
   --cpus boot=1 \
   --log-file /var/cache/modules/vmd/logs/34068-1 \
   --disk path=/mnt/314fdb8d-9dfb-4d7f-9b30-936df0483914/vdisks/34068-1-vda,readonly=off \
   --memory size=2048M \
   --api-socket /var/run/cloud-hypervisor/34068-1 \
   --net tap=t-RMdm4AAbbZLxT
```

**Reply by sboeuf** [reply]
Thanks for sharing that. Any chance you could try to run it without `--initramfs` (by adapting `--cmdline` of course)? I'm trying to understand if the problem might come from this parameter.

**Reply by delandtj** [reply]
hum ... that vm won't boot without initramfs... the whole system lives in it

**Reply by sboeuf** [reply]
Ok I'll do some testing with `--initramfs` on my side.

**Reply by sboeuf** [reply]
@muhamadazmy could you try with `--memory size=4G`?

**Reply by muhamadazmy** [reply]
Just couple of notes:
- This happens randomly (or doesn't)
- Restarting a VM will make the problem go away, it can resume the operation normally. Which makes me doubt it's a problem with the vdisk itself.
- Multiple VMs on the same host can have or not have this issue. So I can't link it to a specific node either :man_shrugging:

**Reply by muhamadazmy** [reply]
> @muhamadazmy could you try with `--memory size=4G`?

I tried with `--memory size=4096M` and I got the same error. I had 2 try couple of times to get the error.

**Reply by sboeuf** [reply]
I'm preparing a fix, I'll let you know when ready and you can have a try :)

**Reply by muhamadazmy** [reply]
Does this mean u managed to reproduce ?

**Reply by sboeuf** [reply]
Could you try the following patch:
```patch
From b1b7d0ac82388a9cb42c9ee6f68d2581e65e5041 Mon Sep 17 00:00:00 2001
From: Sebastien Boeuf <sebastien.boeuf@intel.com>
Date: Fri, 26 Mar 2021 16:50:14 +0100
Subject: [PATCH] fix initramfs

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>
---
 arch/src/x86_64/mod.rs | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/arch/src/x86_64/mod.rs b/arch/src/x86_64/mod.rs
index 3b33cd11..eae542cd 100644
--- a/arch/src/x86_64/mod.rs
+++ b/arch/src/x86_64/mod.rs
@@ -738,15 +738,17 @@ pub fn initramfs_load_addr(
     let first_region = guest_mem
         .find_region(GuestAddress::new(0))
         .ok_or(super::Error::InitramfsAddress)?;
+
     // It's safe to cast to usize because the size of a region can't be greater than usize.
     let lowmem_size = first_region.len() as usize;
 
-    if lowmem_size < initramfs_size {
+    // Let's ensure a 2MiB alignment
+    let alignment: u64 = 1 << 21;
+    if lowmem_size < initramfs_size + alignment as usize {
         return Err(super::Error::InitramfsAddress);
     }
 
-    let aligned_addr: u64 = ((lowmem_size - initramfs_size) & !(crate::pagesize() - 1)) as u64;
-    Ok(aligned_addr)
+    Ok((first_region.last_addr().raw_value() - initramfs_size as u64) & !(alignment - 1))
 }
 
 pub fn get_host_cpu_phys_bits() -> u8 {
-- 
2.27.0


```

**Reply by muhamadazmy** [reply]
Will do, thank you so much for your effort. Will let u know as fast as possible

**Reply by sboeuf** [reply]
FYI we're still investigating as this bug is quite tricky and we're not very fond of the patch here.

**Reply by muhamadazmy** [reply]
@sboeuf unfortunately we still have the same problem 

```
cloud-hypervisor: 183.538173785s: <_disk0_q0> ERROR:block_util/src/lib.rs:198 -- Only head descriptor present: request = Request { request_type: Out, sector: 2306512, data_descriptors: [], status_addr: GuestAddress(0), writeback: true }
cloud-hypervisor: 183.538419411s: <_disk0_q0> ERROR:virtio-devices/src/block.rs:295 -- Failed to process queue (submit): RequestParsing(DescriptorChainTooShort)
```

**Reply by sboeuf** [reply]
@rbradford found the issue and provided the associated fix: https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2450
@muhamadazmy please confirm this is fixing the issue on your side :)

**Reply by muhamadazmy** [reply]
@sboeuf Thank you so much for your amazing and hard work. I confirm that the issue is completely gone from our side, we started many VMs and they all are running well.

Confirming the issue has been fixed.
Thanks again
